### PR TITLE
Add Audio-related classes

### PR DESCRIPTION
**Pull request by Dragonfly911117**
Unlike #5, template code in the App.cpp has been removed.

### CODE REVIEW

**Comment by ntut-rick** [comment]
Why do we need 2 API to play sound

**Comment by Dragonfly911117** [comment]
> Why do we need 2 API to play sound

Cuz Mix_Chunk specifically for handling sound effects, while Mix_Music is intended for background music. The former is well-suited for playing short audio clips. On the other hand, Mix_Music is optimized for seamlessly streaming longer audio tracks, providing efficient decoding and playback of large files.

File: cmake/Dependencies.cmake
[the remark this answers is not in the document]
Comment (Dragonfly911117) [reply cmake/Dependencies.cmake:1]: Cuz CLion didn't like it. Can be undone.

**Comment by ntut-rick** [comment]
> > Why do we need 2 API to play sound
> 
> Cuz Mix_Chunk specifically for handling sound effects, while Mix_Music is intended for background music. The former is well-suited for playing short audio clips. On the other hand, Mix_Music is optimized for seamlessly streaming longer audio tracks, providing efficient decoding and playback of large files.

TODO in new issue #11

File: src/Core/Context.cpp
Comment (kryr34) [inline src/Core/Context.cpp:1]: Is it a good idea to module init/quit sections for easy management (or testing) ?
> maybe open another issue and let a future one handle it

File: src/Util/BGM.cpp
Comment (kryr34) [inline src/Util/BGM.cpp:1]: I don't think `assert()` is a good idea to handle uninitialized. Maybe leave `FIXME`/`TODO`?

File: include/Util/BGM.hpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * @brief Increases the volume of the background music by one.
+     */
+    void VolumeUp();
+
+    /**
+     * @brief Decreases the volume of the background music by one.
+     */
+    void VolumeDown();
```
Comment (kryr34) [inline include/Util/BGM.hpp:70]: Looks not useful, how about merge two function and add `step` argument?

File: include/Util/BGM.hpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    void Play(const int &loop = -1);
+
+    /**
+     * @brief Fades in the background music gradually.
+     * @param tick The duration of the fade-in effect, in milliseconds.
+     * @param loop The number of times the music will loop after the fade-in is
+     *                      complete.<br> A value of -1 means it will loop indefinitely.
+     */
+    void FadeIn(const int &tick, const int &loop = -1);
```
Comment (kryr34) [inline include/Util/BGM.hpp:88]: Why no fade out
https://wiki.libsdl.org/SDL2_mixer/Mix_FadeOutMusic

**Review by kryr34 — CHANGES_REQUESTED** [review]

File: include/Util/BGM.hpp (hunk 1 above)
Comment (ntut-rick) [inline include/Util/BGM.hpp:70]: Looks not useful, how about merge two function and add `step` argument?

File: include/Util/BGM.hpp (hunk 2 above)
Comment (ntut-rick) [inline include/Util/BGM.hpp:88]: Why no fade out
https://wiki.libsdl.org/SDL2_mixer/Mix_FadeOutMusic

File: src/Core/Context.cpp
Comment (ntut-rick) [inline src/Core/Context.cpp:1]: Is it a good idea to module init/quit sections for easy management (or testing) ?
> maybe open another issue and let a future one handle it

File: src/Util/BGM.cpp
Comment (ntut-rick) [inline src/Util/BGM.cpp:1]: I don't think `assert()` is a good idea to handle uninitialized. Maybe leave `FIXME`/`TODO`?

**Review by ntut-rick — CHANGES_REQUESTED** [review]

File: src/Core/Context.cpp
Comment (Dragonfly911117) [reply src/Core/Context.cpp:1]: I think it is better to init/quit with other SDL-things thought I get the point that this is audio side not video.

File: test/audio.cpp
Comment (NOOBDY) [inline test/audio.cpp:1]: There are a lot of wrong naming case and missing line break here

File: test/audio.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	auto yn = std::cin.get();
+	std::cin.clear();
+	return yn != 'n';
+}
+void EXCEPT_USER_SAY_YES(const std::string text) {
+	if (user_check(text)) {
+		/* pass */
+	} else {
+		ADD_FAILURE() << "user test: ["+text+"] is false";
+	}
```
Comment (NOOBDY) [inline test/audio.cpp:18]:
```suggestion
	if (!user_check(text)) {
		ADD_FAILURE() << "user test: ["+text+"] is false";
	}
```

File: test/audio.cpp
Comment (NOOBDY) [inline test/audio.cpp:1]: `std::string` in functions should be passed in as `const std::string &parameter`

**Review by NOOBDY — CHANGES_REQUESTED** [review]

File: include/Util/BGM.hpp (hunk 1 above)
Comment (Dragonfly911117) [reply include/Util/BGM.hpp]: Done

File: include/Util/BGM.hpp (hunk 2 above)
Comment (Dragonfly911117) [reply include/Util/BGM.hpp]: Done

File: CMakeLists.txt [hunk 4]
```diff
@@ -64,7 +69,7 @@ target_include_directories(Sample SYSTEM PRIVATE
     ${DEPENDENCY_INCLUDE_DIRS}
 )
 target_include_directories(Sample PRIVATE
-    ${CMAKE_SOURCE_DIR}/include
+    ${INCLUDE_DIR}
```
[the remark this answers is not in the document]
Comment (ntut-rick) [reply CMakeLists.txt:72]: The variable named `INCLUDE_DIR` doesn't mean it's a directory?

File: src/Util/SFX.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#include "Util/SFX.hpp"
+#include "Util/Logger.hpp"
+
+namespace Util {
+
+SFX::SFX(const std::string &path) {
+    m_Chunk = std::unique_ptr<Mix_Chunk, void (*)(Mix_Chunk *)>(
+        Mix_LoadWAV(path.c_str()), Mix_FreeChunk);
+    if (m_Chunk == nullptr) {
+        LOG_DEBUG("Failed to load SFX: " + path + std::string(Mix_GetError()));
```
Comment (NOOBDY) [inline src/Util/SFX.cpp:10]:
```suggestion
        LOG_DEBUG("Failed to load SFX: {}{}", path, std::string(Mix_GetError()));
```

**Review by NOOBDY — CHANGES_REQUESTED** [review]
no need to pass int as `const int &` as it's not faster
https://stackoverflow.com/a/3009568

**Review by NOOBDY — APPROVED** [review]
LGTM

**Review by ntut-rick — APPROVED** [review]

**Pull request merged**